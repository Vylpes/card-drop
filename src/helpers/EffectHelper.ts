import {EmbedBuilder} from "discord.js";
import UserEffect from "../database/entities/app/UserEffect";
import EmbedColours from "../constants/EmbedColours";

export default class EffectHelper {
    public static async AddEffectToUserInventory(userId: string, name: string, quantity: number = 1) {
        let effect = await UserEffect.FetchOneByUserIdAndName(userId, name);

        if (!effect) {
            effect = new UserEffect(name, userId, quantity);
        } else {
            effect.AddUnused(quantity);
        }

        await effect.Save(UserEffect, effect);
    }

    public static async UseEffect(userId: string, name: string, whenExpires: Date): Promise<boolean> {
        const effect = await UserEffect.FetchOneByUserIdAndName(userId, name);
        const now = new Date();

        if (!effect || effect.Unused == 0) {
            return false;
        }

        if (effect.WhenExpires && now < effect.WhenExpires) {
            return false;
        }

        effect.UseEffect(whenExpires);

        await effect.Save(UserEffect, effect);

        return true;
    }

    public static async HasEffect(userId: string, name: string): Promise<boolean> {
        const effect = await UserEffect.FetchOneByUserIdAndName(userId, name);
        const now = new Date();

        if (!effect || !effect.WhenExpires) {
            return false;
        }

        if (now > effect.WhenExpires) {
            return false;
        }

        return true;
    }

    public static async GenerateEffectEmbed(userId: string, page: number): Promise<EmbedBuilder> {
        const effects = await UserEffect.FetchAllByUserIdPaginated(userId, page - 1);

        let description = "*none*";

        if (effects.length > 0) {
            description = effects.map(x => `${x.Name} x${x.Unused}`).join("\n");
        }

        const embed = new EmbedBuilder()
            .setTitle("Effects")
            .setDescription(description)
            .setColor(EmbedColours.Ok);

        return embed;
    }
}
