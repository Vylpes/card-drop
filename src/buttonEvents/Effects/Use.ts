import { ActionRowBuilder, ButtonBuilder, ButtonInteraction, ButtonStyle, EmbedBuilder } from "discord.js";
import { EffectDetails } from "../../constants/EffectDetails";
import EffectHelper from "../../helpers/EffectHelper";
import EmbedColours from "../../constants/EmbedColours";
import TimeLengthInput from "../../helpers/TimeLengthInput";
import AppLogger from "../../client/appLogger";

export default async function Use(interaction: ButtonInteraction) {
    const subaction = interaction.customId.split(" ")[2];

    switch (subaction) {
        case "confirm":
            await UseConfirm(interaction);
            break;
        case "cancel":
            await UseCancel(interaction);
            break;
    }
}

export async function UseConfirm(interaction: ButtonInteraction) {
    const id = interaction.customId.split(" ")[3];

    const effectDetail = EffectDetails.get(id);

    if (!effectDetail) {
        AppLogger.LogError("Button/Effects/Use", `Effect not found, ${id}`);

        await interaction.reply("Effect not found in system!");
        return;
    }

    const now = new Date();

    const whenExpires = new Date(now.getTime() + effectDetail.duration);

    const result = await EffectHelper.UseEffect(interaction.user.id, id, whenExpires);

    if (!result) {
        await interaction.reply("Unable to use effect! Please make sure you have it in your inventory and is not on cooldown");
        return;
    }

    const embed = new EmbedBuilder()
        .setTitle("Effect Used")
        .setDescription("You now have an active effect!")
        .setColor(EmbedColours.Green)
        .addFields([
            {
                name: "Effect",
                value: effectDetail.friendlyName,
                inline: true,
            },
            {
                name: "Expires",
                value: `<t:${Math.round(whenExpires.getTime() / 1000)}:f>`,
                inline: true,
            },
        ]);

    const row = new ActionRowBuilder<ButtonBuilder>()
        .addComponents([
            new ButtonBuilder()
                .setLabel("Confirm")
                .setCustomId(`effects use confirm ${effectDetail.id}`)
                .setStyle(ButtonStyle.Primary)
                .setDisabled(true),
            new ButtonBuilder()
                .setLabel("Cancel")
                .setCustomId(`effects use cancel ${effectDetail.id}`)
                .setStyle(ButtonStyle.Danger)
                .setDisabled(true),
        ]);

    await interaction.update({
        embeds: [ embed ],
        components: [ row ],
    });
}

export async function UseCancel(interaction: ButtonInteraction) {
    const id = interaction.customId.split(" ")[3];

    const effectDetail = EffectDetails.get(id);

    if (!effectDetail) {
        AppLogger.LogError("Button/Effects/Cancel", `Effect not found, ${id}`);

        await interaction.reply("Effect not found in system!");
        return;
    }

    const timeLengthInput = TimeLengthInput.ConvertFromMilliseconds(effectDetail.duration);

    const embed = new EmbedBuilder()
        .setTitle("Effect Use Cancelled")
        .setDescription("The effect from your inventory has not been used")
        .setColor(EmbedColours.Grey)
        .addFields([
            {
                name: "Effect",
                value: effectDetail.friendlyName,
                inline: true,
            },
            {
                name: "Expires",
                value: timeLengthInput.GetLengthShort(),
                inline: true,
            },
        ]);

    const row = new ActionRowBuilder<ButtonBuilder>()
        .addComponents([
            new ButtonBuilder()
                .setLabel("Confirm")
                .setCustomId(`effects use confirm ${effectDetail.id}`)
                .setStyle(ButtonStyle.Primary)
                .setDisabled(true),
            new ButtonBuilder()
                .setLabel("Cancel")
                .setCustomId(`effects use cancel ${effectDetail.id}`)
                .setStyle(ButtonStyle.Danger)
                .setDisabled(true),
        ]);

    await interaction.update({
        embeds: [ embed ],
        components: [ row ],
    });
}