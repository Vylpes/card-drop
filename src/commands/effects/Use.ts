import { ActionRowBuilder, ButtonBuilder, ButtonStyle, CommandInteraction, EmbedBuilder } from "discord.js";
import { EffectDetails } from "../../constants/EffectDetails";
import AppLogger from "../../client/appLogger";
import EffectHelper from "../../helpers/EffectHelper";
import TimeLengthInput from "../../helpers/TimeLengthInput";
import EmbedColours from "../../constants/EmbedColours";

export default async function Use(interaction: CommandInteraction) {
    const id = interaction.options.get("id", true).value!.toString();

    const effectDetail = EffectDetails.get(id);

    if (!effectDetail) {
        AppLogger.LogWarn("Commands/Effects", `Unable to find effect details for ${id}`);

        await interaction.reply("Unable to find effect!");
        return;
    }

    const canUseEffect = await EffectHelper.CanUseEffect(interaction.user.id, id);

    if (!canUseEffect) {
        await interaction.reply("Unable to use effect! Please make sure you have it in your inventory and is not on cooldown");
        return;
    }

    const timeLengthInput = TimeLengthInput.ConvertFromMilliseconds(effectDetail.duration);

    const embed = new EmbedBuilder()
        .setTitle("Effect Confirmation")
        .setDescription("Would you like to use this effect?")
        .setColor(EmbedColours.Ok)
        .addFields([
            {
                name: "Effect",
                value: effectDetail.friendlyName,
                inline: true,
            },
            {
                name: "Length",
                value: timeLengthInput.GetLengthShort(),
                inline: true,
            },
        ]);

    const row = new ActionRowBuilder<ButtonBuilder>()
        .addComponents([
            new ButtonBuilder()
                .setLabel("Confirm")
                .setCustomId(`effects use confirm ${effectDetail.id}`)
                .setStyle(ButtonStyle.Primary),
            new ButtonBuilder()
                .setLabel("Cancel")
                .setCustomId(`effects use cancel ${effectDetail.id}`)
                .setStyle(ButtonStyle.Danger),
        ]);

    await interaction.reply({
        embeds: [ embed ],
        components: [ row ],
    });
}