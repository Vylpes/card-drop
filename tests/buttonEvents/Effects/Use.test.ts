import { ButtonInteraction, EmbedBuilder, InteractionResponse } from "discord.js";
import Use from "../../../src/buttonEvents/Effects/Use";
import { mock } from "jest-mock-extended";
import AppLogger from "../../../src/client/appLogger";
import EffectHelper from "../../../src/helpers/EffectHelper";

jest.mock("../../../src/client/appLogger");
jest.mock("../../../src/helpers/EffectHelper");

beforeEach(() => {
    jest.resetAllMocks();

    jest.useFakeTimers();
    jest.setSystemTime(0);
});

afterAll(() => {
    jest.useRealTimers();
});

describe("Execute", () => {
    test("GIVEN subaction is unknown, EXPECT nothing to be called", async () => {
        // Arrange
        const interaction = mock<ButtonInteraction>();
        interaction.customId = "effects use invalud";

        // Act
        await Use.Execute(interaction);

        // Assert
        expect(interaction.reply).not.toHaveBeenCalled();
        expect(interaction.update).not.toHaveBeenCalled();
    });
});

describe("UseConfirm", () => {
    let interaction = mock<ButtonInteraction>();

    beforeEach(() => {
        interaction = mock<ButtonInteraction>();
        interaction.customId = "effects use confirm";
    });

    test("GIVEN effectDetail is not found, EXPECT error", async () => {
        // Arrange
        interaction.customId += " invalid";

        // Act
        await Use.Execute(interaction);

        // Assert
        expect(AppLogger.LogError).toHaveBeenCalledTimes(1);
        expect(AppLogger.LogError).toHaveBeenCalledWith("Button/Effects/Use", "Effect not found, invalid");

        expect(interaction.reply).toHaveBeenCalledTimes(1);
        expect(interaction.reply).toHaveBeenCalledWith("Effect not found in system!");
    });

    test("GIVEN EffectHelper.UseEffect failed, EXPECT error", async () => {
        // Arrange
        interaction.customId += " unclaimed";
        interaction.user.id = "userId";

        (EffectHelper.UseEffect as jest.Mock).mockResolvedValue(false);

        const whenExpires = new Date(Date.now() + 10 * 60 * 1000);

        // Act
        await Use.Execute(interaction);

        // Assert
        expect(EffectHelper.UseEffect).toHaveBeenCalledTimes(1);
        expect(EffectHelper.UseEffect).toHaveBeenCalledWith("userId", "unclaimed", whenExpires);

        expect(interaction.reply).toHaveBeenCalledTimes(1);
        expect(interaction.reply).toHaveBeenCalledWith("Unable to use effect! Please make sure you have it in your inventory and is not on cooldown");
    });

    test("GIVEN EffectHelper.UseEffect succeeded, EXPECT interaction updated", async () => {
        let updatedWith;

        // Arrange
        interaction.customId += " unclaimed";
        interaction.user.id = "userId";
        interaction.update.mockImplementation(async (opts: any) => {
            updatedWith = opts;

            return mock<InteractionResponse<boolean>>();
        });

        (EffectHelper.UseEffect as jest.Mock).mockResolvedValue(true);

        const whenExpires = new Date(Date.now() + 10 * 60 * 1000);

        // Act
        await Use.Execute(interaction);

        // Assert
        expect(EffectHelper.UseEffect).toHaveBeenCalledTimes(1);
        expect(EffectHelper.UseEffect).toHaveBeenCalledWith("userId", "unclaimed", whenExpires);

        expect(interaction.update).toHaveBeenCalledTimes(1);
        expect(updatedWith).toMatchSnapshot();
    });
});

describe("UseCancel", () => {
    let interaction = mock<ButtonInteraction>();

    beforeEach(() => {
        interaction = mock<ButtonInteraction>();
        interaction.customId = "effects use cancel";
    });

    test("GIVEN effectDetail is not found, EXPECT error", async () => {
        // Arrange
        interaction.customId += " invalid";

        // Act
        await Use.Execute(interaction);

        // Assert
        expect(AppLogger.LogError).toHaveBeenCalledTimes(1);
        expect(AppLogger.LogError).toHaveBeenCalledWith("Button/Effects/Cancel", "Effect not found, invalid");

        expect(interaction.reply).toHaveBeenCalledTimes(1);
        expect(interaction.reply).toHaveBeenCalledWith("Effect not found in system!");
    });

    test("GIVEN effectDetail is found, EXPECT interaction updated", async () => {
        let updatedWith;

        // Arrange
        interaction.customId += " unclaimed";
        interaction.user.id = "userId";
        interaction.update.mockImplementation(async (opts: any) => {
            updatedWith = opts;

            return mock<InteractionResponse<boolean>>();
        });
        // Act
        await Use.Execute(interaction);

        // Assert
        expect(interaction.update).toHaveBeenCalledTimes(1);
        expect(updatedWith).toMatchSnapshot();
    });
});