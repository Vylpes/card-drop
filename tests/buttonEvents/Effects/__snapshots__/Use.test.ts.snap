// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UseCancel GIVEN effectDetail is found, EXPECT interaction updated 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "effects use confirm unclaimed",
          "disabled": true,
          "emoji": undefined,
          "label": "Confirm",
          "style": 1,
          "type": 2,
        },
        {
          "custom_id": "effects use cancel unclaimed",
          "disabled": true,
          "emoji": undefined,
          "label": "Cancel",
          "style": 4,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "embeds": [
    {
      "color": 13882323,
      "description": "The effect from your inventory has not been used",
      "fields": [
        {
          "inline": true,
          "name": "Effect",
          "value": "Unclaimed Chance Up",
        },
        {
          "inline": true,
          "name": "Expires",
          "value": "10m",
        },
      ],
      "title": "Effect Use Cancelled",
    },
  ],
}
`;

exports[`UseConfirm GIVEN EffectHelper.UseEffect succeeded, EXPECT interaction updated 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "effects use confirm unclaimed",
          "disabled": true,
          "emoji": undefined,
          "label": "Confirm",
          "style": 1,
          "type": 2,
        },
        {
          "custom_id": "effects use cancel unclaimed",
          "disabled": true,
          "emoji": undefined,
          "label": "Cancel",
          "style": 4,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "embeds": [
    {
      "color": 2263842,
      "description": "You now have an active effect!",
      "fields": [
        {
          "inline": true,
          "name": "Effect",
          "value": "Unclaimed Chance Up",
        },
        {
          "inline": true,
          "name": "Expires",
          "value": "<t:600:f>",
        },
      ],
      "title": "Effect Used",
    },
  ],
}
`;
